File: src/main/java/com/jozufozu/exnihiloomnia/common/registries/JsonHelper.java
package com.jozufozu.exnihiloomnia.common.registries;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSyntaxException;
import com.jozufozu.exnihiloomnia.common.lib.RegistriesLib;
import com.jozufozu.exnihiloomnia.common.registries.ingredients.IngredientNBT;
import com.jozufozu.exnihiloomnia.common.util.Color;
import net.minecraft.block.Block;
import net.minecraft.init.Blocks;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.Ingredient;
import net.minecraft.nbt.JsonToNBT;
import net.minecraft.nbt.NBTException;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.JsonUtils;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.oredict.OreDictionary;
import net.minecraftforge.oredict.OreIngredient;

import java.util.function.Predicate;

public class JsonHelper
{
    public static Ingredient deserializeBlockIngredient(JsonObject object) throws JsonParseException
    {
        return deserializeIngredient(object, item -> Block.getBlockFromItem(item.getItem()) != Blocks.AIR);
    }
    
    public static Ingredient deserializeIngredient(JsonObject object, Predicate<ItemStack> requirements) throws JsonParseException
    {
        Ingredient out = deserializeIngredient(object);
        
        boolean meetsReq = false;
    
        for (ItemStack itemStack : out.getMatchingStacks())
        {
            if (requirements.test(itemStack))
                meetsReq = true;
        }
        if (!meetsReq)
            throw new JsonSyntaxException("Given input " + object.toString() + "' is invalid!");
        
        return out;
    }
    
    /**
     * Makes an {@link Ingredient} object from Json. Will try to find an oredict entry from an item ID if the oredict tag has an item resource location
     */
    public static Ingredient deserializeIngredient(JsonObject input) throws JsonParseException
    {
        boolean isItem = input.has(RegistriesLib.ITEM_ID);
        boolean isOredict = input.has(RegistriesLib.OREDICT);
        boolean isNbt = input.has(RegistriesLib.NBT);
        
        if (isItem && isOredict)
        {
            throw new JsonSyntaxException("Cannot have both oredict and item tags in input!");
        }
    
        if (isNbt && isOredict)
        {
            throw new JsonSyntaxException("Cannot have both oredict and nbt tags in input!");
        }
        
        if (isItem)
        {
            ItemStack itemStack = deserializeItem(input, false);
            
            if (!input.has(RegistriesLib.DATA))
                itemStack = new ItemStack(itemStack.getItem(), 1, OreDictionary.WILDCARD_VALUE);
    
            if (isNbt)
            {
                return new IngredientNBT(itemStack);
            }
            
            return Ingredient.fromStacks(itemStack);
        }
        else if (isOredict)
        {
            String oreDict = JsonUtils.getString(input, RegistriesLib.OREDICT);
            
            if (oreDict.contains(":"))
            {
                Item item = Item.REGISTRY.getObject(new ResourceLocation(oreDict));
                
                if (item == null)
                {
                    throw new JsonSyntaxException("Unknown item '" + oreDict + "'");
                }
                
                ItemStack stack = new ItemStack(item, 1, OreDictionary.WILDCARD_VALUE);
                int[] oreIDs = OreDictionary.getOreIDs(stack);
                
                if (oreIDs.length == 0)
                {
                    RegistryLoader.error("Given item ingredient, '" + oreDict + "' has no Ore Dictionary entries! Using the item instead.");
                    
                    return Ingredient.fromStacks(stack);
                }
                
                return new OreIngredient(OreDictionary.getOreName(oreIDs[0]));
            }
            
            return new OreIngredient(oreDict);
        }
        
        throw new JsonSyntaxException("Recipe is missing input!");
    }
    
    /**
     * Turns a {@link JsonObject} into an {@link ItemStack} using the standard id, data, count tags.
     * @param useCount Whether or not to use the provided count tag. ItemStack will have 1 item if false.
     */
    public static ItemStack deserializeItem(JsonObject itemObject, boolean useCount) throws JsonParseException
    {
        String resourceName = JsonUtils.getString(itemObject, RegistriesLib.ITEM_ID);
        Item item = Item.REGISTRY.getObject(new ResourceLocation(resourceName));
        
        if (item == null)
        {
            throw new JsonSyntaxException("Unknown item '" + resourceName + "'");
        }

        int data = JsonUtils.getInt(itemObject, RegistriesLib.DATA, 0);
        int count = useCount ? JsonUtils.getInt(itemObject, RegistriesLib.COUNT, 1) : 1;
    
        if (itemObject.has("nbt"))
        {
            try
            {
                JsonElement element = itemObject.get("nbt");
                NBTTagCompound nbt;
                if(element.isJsonObject())
                    nbt = JsonToNBT.getTagFromJson(RegistryLoader.GSON.toJson(element));
                else
                    nbt = JsonToNBT.getTagFromJson(element.getAsString());
            
                NBTTagCompound tmp = new NBTTagCompound();
                if (nbt.hasKey("ForgeCaps"))
                {
                    tmp.setTag("ForgeCaps", nbt.getTag("ForgeCaps"));
                    nbt.removeTag("ForgeCaps");
                }
            
                tmp.setTag("tag", nbt);
                tmp.setString("id", resourceName);
                tmp.setInteger("Count", count);
                tmp.setInteger("Damage", data);
            
                return new ItemStack(tmp);
            }
            catch (NBTException e)
            {
                throw new JsonSyntaxException("Invalid NBT Entry: " + e.toString());
            }
        }
        
        return new ItemStack(item, count, data);
    }
    
//    public static Fluid deserializeFluid(JsonObject object, String memberName) throws JsonParseException
//    {
//        String name = JsonUtils.getString(object, memberName);
//
//        if (!FluidRegistry.isFluidRegistered(name))
//        {
//            throw new JsonSyntaxException("Fluid '" + name + "' does not exist!");
//        }
//
//        return FluidRegistry.getFluid(name);
//    }
    
    /**
     * Turns a given string in either hex or r, g, b(, a) format into a packed color int
     */
    public static Color deserializeColor(String colorString) throws JsonParseException
    {
        int color = 0;
        
        if (colorString.contains(","))
        {
            colorString = colorString.replace(" ", "");
            String[] rgba = colorString.split(",");
            
            if (rgba.length > 4 || rgba.length < 3)
                throw new JsonSyntaxException("Could not parse color '" + colorString + "'");
            
            color |= Integer.parseInt(rgba[0]) << 16;
            color |= Integer.parseInt(rgba[1]) << 8;
            color |= Integer.parseInt(rgba[2]);
    
            boolean alpha = rgba.length == 4;
            if (alpha)
            {
                color |= Integer.parseInt(rgba[3]) << 24;
            }
            
            return new Color(color, alpha);
        }
        else
        {
            if (colorString.length() == 8 || colorString.length() == 6)
            {
                color = Integer.parseInt(colorString, 16);
    
                return new Color(color, colorString.length() == 8);
            }
        }
        
        throw new JsonSyntaxException("Could not parse color '" + colorString + "'");
    }
}

File: src/main/java/com/jozufozu/exnihiloomnia/common/registries/JsonHelper.kt
package com.jozufozu.exnihiloomnia.common.registries

import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonSyntaxException
import com.jozufozu.exnihiloomnia.common.lib.LibRegistries
import com.jozufozu.exnihiloomnia.common.registries.ingredients.IngredientNBT
import com.jozufozu.exnihiloomnia.common.util.Color
import net.minecraft.block.Block
import net.minecraft.init.Blocks
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.Ingredient
import net.minecraft.nbt.JsonToNBT
import net.minecraft.nbt.NBTException
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.JsonUtils
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fluids.Fluid
import net.minecraftforge.fluids.FluidRegistry
import net.minecraftforge.oredict.OreDictionary
import net.minecraftforge.oredict.OreIngredient

object JsonHelper {
    @Throws(JsonParseException::class)
    fun deserializeBlockIngredient(jsonObject: JsonObject): Ingredient {
        return deserializeIngredient(jsonObject) { Block.getBlockFromItem(it.item) !== Blocks.AIR }
    }

    @Throws(JsonParseException::class)
    fun deserializeIngredient(jsonObject: JsonObject, requirements: (ItemStack) -> Boolean): Ingredient {
        val out = deserializeIngredient(jsonObject)

        var meetsReq = false

        for (itemStack in out.matchingStacks) {
            if (requirements(itemStack))
                meetsReq = true
        }
        if (!meetsReq)
            throw JsonSyntaxException("Given input $jsonObject' is invalid!")

        return out
    }

    /**
     * Makes an [Ingredient] object from Json. Will try to find an oredict entry from an item ID if the oredict tag has an item resource location
     */
    @Throws(JsonParseException::class)
    fun deserializeIngredient(input: JsonObject): Ingredient {
        val isItem = input.has(LibRegistries.ID)
        val isOredict = input.has(LibRegistries.OREDICT)
        val isNbt = input.has(LibRegistries.NBT)

        if (isItem && isOredict) {
            throw JsonSyntaxException("Cannot have both oredict and item tags in input!")
        }

        if (isNbt && isOredict) {
            throw JsonSyntaxException("Cannot have both oredict and nbt tags in input!")
        }

        if (isItem) {
            var itemStack = deserializeItem(input, false)

            if (!input.has(LibRegistries.DATA))
                itemStack = ItemStack(itemStack.item, 1, OreDictionary.WILDCARD_VALUE)

            return if (isNbt) {
                IngredientNBT(itemStack)
            } else Ingredient.fromStacks(itemStack)

        } else if (isOredict) {
            val oreDict = JsonUtils.getString(input, LibRegistries.OREDICT)

            if (oreDict.contains(":")) {
                val item = Item.REGISTRY.getObject(ResourceLocation(oreDict))
                        ?: throw JsonSyntaxException("Unknown item '$oreDict'")

                val stack = ItemStack(item, 1, OreDictionary.WILDCARD_VALUE)
                val oreIDs = OreDictionary.getOreIDs(stack)

                if (oreIDs.isEmpty()) {
                    RegistryLoader.error("Given item ingredient, '$oreDict' has no Ore Dictionary entries! Using the item instead.")

                    return Ingredient.fromStacks(stack)
                }

                return OreIngredient(OreDictionary.getOreName(oreIDs[0]))
            }

            return OreIngredient(oreDict)
        }

        throw JsonSyntaxException("Recipe is missing input!")
    }

    /**
     * Turns a [JsonObject] into an [ItemStack] using the standard id, data, count tags.
     * @param useCount Whether or not to use the provided count tag. ItemStack will have 1 item if false.
     */
    @Throws(JsonParseException::class)
    fun deserializeItem(itemObject: JsonObject, useCount: Boolean): ItemStack {
        val resourceName = JsonUtils.getString(itemObject, LibRegistries.ID)
        val item = Item.REGISTRY.getObject(ResourceLocation(resourceName))
                ?: throw JsonSyntaxException("Unknown item '$resourceName'")

        val data = JsonUtils.getInt(itemObject, LibRegistries.DATA, 0)
        val count = if (useCount) JsonUtils.getInt(itemObject, LibRegistries.COUNT, 1) else 1

        if (itemObject.has("nbt")) {
            try {
                val element = itemObject.get("nbt")
                val nbt: NBTTagCompound = if (element.isJsonObject)
                    JsonToNBT.getTagFromJson(RegistryLoader.GSON.toJson(element))
                else
                    JsonToNBT.getTagFromJson(element.asString)

                val tmp = NBTTagCompound()
                if (nbt.hasKey("ForgeCaps")) {
                    tmp.setTag("ForgeCaps", nbt.getTag("ForgeCaps"))
                    nbt.removeTag("ForgeCaps")
                }

                tmp.setTag("tag", nbt)
                tmp.setString("id", resourceName)
                tmp.setInteger("Count", count)
                tmp.setInteger("Damage", data)

                return ItemStack(tmp)
            } catch (e: NBTException) {
                throw JsonSyntaxException("Invalid NBT Entry: $e")
            }

        }

        return ItemStack(item, count, data)
    }

    @Throws(JsonParseException::class)
    fun deserializeFluid(jsonObject: JsonObject, memberName: String): Fluid {
        val name = JsonUtils.getString(jsonObject, memberName)

        if (!FluidRegistry.isFluidRegistered(name)) {
            throw JsonSyntaxException("Fluid '$name' does not exist!")
        }

        return FluidRegistry.getFluid(name)
    }

    /**
     * Turns a given string in either hex or r, g, b(, a) format into a packed color int
     */
    @Throws(JsonParseException::class)
    fun deserializeColor(colorString: String): Color {
        var colorString = colorString
        var color = 0

        if (colorString.contains(",")) {
            colorString = colorString.replace(" ", "")
            val rgba = colorString.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            if (rgba.size > 4 || rgba.size < 3)
                throw JsonSyntaxException("Could not parse color '$colorString'")

            color = color or (Integer.parseInt(rgba[0]) shl 16)
            color = color or (Integer.parseInt(rgba[1]) shl 8)
            color = color or Integer.parseInt(rgba[2])

            val alpha = rgba.size == 4
            if (alpha) {
                color = color or (Integer.parseInt(rgba[3]) shl 24)
            }

            return Color(color, alpha)
        } else {
            if (colorString.length == 8 || colorString.length == 6) {
                color = Integer.parseInt(colorString, 16)

                return Color(color, colorString.length == 8)
            }
        }

        throw JsonSyntaxException("Could not parse color '$colorString'")
    }
}
